instrumentation/instaredis/instaredis_test.go

```diff
@@ -0,0 +1,347 @@
+// (c) Copyright IBM Corp. 2022
+
+//go:build go1.13
+// +build go1.13
+
+package instaredis_test
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/go-redis/redis/v8"
+	instana "github.com/instana/go-sensor"
+	"github.com/instana/go-sensor/instrumentation/instaredis"
+	"github.com/instana/testify/assert"
+	"github.com/instana/testify/require"
+)
+
+/*
+Redis small cheat sheet (command sent vs characters received):
+
+SET  = +OK\r\n
+GET  = $4\r\n Where 4 is the size of the item "gotten"
+DEL  = :1\r\n Where 1 is the amount of data deleted (I guess)
+INCR = :7\r\n Where 7 is the new value of the variable incremented
+*/
+
+type redisType int
+
+const (
+	Single redisType = iota
+	SingleFailover
+	Cluster
+	ClusterFailover
+)
+
+var redisTypeMap = map[redisType]string{
+	Single:          "Single Server",
+	SingleFailover:  "Single Failover Server",
+	Cluster:         "Cluster Server",
+	ClusterFailover: "Cluster Failover Server",
+}
+
+// returns an instance of redis.Client, compatible with redis.NewClient() and redis.NewFailoverClient()
+func buildNewClient(hasSentinel bool) *redis.Client {
+	if hasSentinel {
+		return redis.NewFailoverClient(&redis.FailoverOptions{
+			SlaveOnly:     true,
+			RouteRandomly: false,
+			MasterName:    "redis1",
+			MaxRetries:    1,
+			SentinelAddrs: []string{":26379"},
+			Dialer: func(ctx context.Context, network string, addr string) (net.Conn, error) {
+				netConn := getMockConn(SingleFailover, ctx, network, addr)
+				return netConn, nil
+			},
+		})
+	} else {
+		return redis.NewClient(&redis.Options{
+			Addr: ":6382",
+			Dialer: func(ctx context.Context, network string, addr string) (net.Conn, error) {
+				netConn := getMockConn(Single, ctx, network, addr)
+				return netConn, nil
+			},
+		})
+	}
+}
+
+// returns an instance of redis.ClusterClient, compatible with redis.NewClusterClient() and redis.NewFailoverClusterClient()
+func buildNewClusterClient(hasSentinel bool) *redis.ClusterClient {
+	if hasSentinel {
+		return redis.NewFailoverClusterClient(&redis.FailoverOptions{
+			MasterName:    "redis1",
+			MaxRetries:    1,
+			SentinelAddrs: []string{":26379"},
+			Dialer: func(ctx context.Context, network string, addr string) (net.Conn, error) {
+				netConn := getMockConn(ClusterFailover, ctx, network, addr)
+				return netConn, nil
+			},
+		})
+	} else {
+		return redis.NewClusterClient(&redis.ClusterOptions{
+			Addrs: []string{":6382"},
+			Dialer: func(ctx context.Context, network string, addr string) (net.Conn, error) {
+				netConn := getMockConn(Cluster, ctx, network, addr)
+				return netConn, nil
+			},
+		})
+	}
+}
+
+type MockAddr struct {
+	TheNetwork string
+	TheAddr    string
```

Comment:
I'd suggest to not to export fields that have getters anyway.
```suggestion
	network string
	addr    string
```